include/drivers/rtio.h

```diff
@@ -0,0 +1,482 @@
+			(struct rtio_mempool_block *)(memblock.data);
+		u8_t *dataptr = (u8_t *)memblock.data + sizeof(*mempool_block);
+
+		mempool_block->id = memblock.id;
+		*block = (struct rtio_block *)mempool_block;
+		rtio_block_init(*block, dataptr, size);
+	}
+	return res;
+}
+
+static inline void rtio_mempool_block_free(
+	struct rtio_block_allocator *allocator,
+	struct rtio_block *block)
+{
+	struct rtio_mempool_block *mempool_block =
+		(struct rtio_mempool_block *)block;
+
+	k_mem_pool_free_id(&mempool_block->id);
+}
+
+/**
+ * @brief Define an rtio block mempool allocator
+ */
+#define RTIO_MEMPOOL_ALLOCATOR_DEFINE(_name, _minsz, _maxsz, _nmax)	\
+	K_MEM_POOL_DEFINE(rtio_pool_mempool_##_name,			\
+			  sizeof(struct rtio_mempool_block) + _minsz,	\
+			  sizeof(struct rtio_mempool_block) + _maxsz,	\
+			  _nmax,					\
+			  4);						\
+									\
+	struct rtio_mempool_block_allocator _name = {			\
+	.allocator = {							\
+	.alloc = rtio_mempool_block_alloc,				\
+	.free = rtio_mempool_block_free					\
+},									\
+	.mempool = &rtio_pool_mempool_##_name				\
+}
+
+
+/**
+ * @brief Allocate a rtio_block from an rtio_pool with a timeout
+ *
+ * This call is not safe to do with a timeout other than K_NO_WAIT
+ * in an interrupt handler.
+ *
+ * The allocator *must* have a memory layout equivalent to
+ * struct rtio_block_allocator
+ */
+static inline int rtio_block_alloc(void *allocatorv,
+				   struct rtio_block **block,
+				   size_t size,
+				   u32_t timeout)
+{
+	struct rtio_block_allocator *allocator =
+		(struct rtio_block_allocator *)allocatorv;
+
+	return allocator->alloc(allocator, block, size, timeout);
+
+}
+
+/**
+ * @brief Free a rtio_block back to the allocator
+ */
+static inline void rtio_block_free(void *allocatorv, struct rtio_block *block)
+{
+	struct rtio_block_allocator *allocator =
+		(struct rtio_block_allocator *)allocatorv;
+	allocator->free(allocator, block);
+}
+
+/**
+ * @brief Output config describing where and when to push a blocks
+ *
+ * Output config should at least have a valid allocator and k_fifo pointer given
+ *
+ * Drivers should make a best attempt at fulfilling the policy of when
+ * to return and notify the caller the block is ready by pushing it
+ * into the rtio_pipe.
+ *
+ * If the buffer is full before the time or the desired number of
+ * items is fulfilled its likely an error on the authors part and should
+ * be logged as a warning when possible. This is a buffer overrun
+ * scenario on the reader side.
+ *
+ * The behavior when the time is K_FOREVER and number is 0 the buffer
+ * will be marked completed immediately and no read or write will take place.
+ */
+struct rtio_output_config {
+	/**
+	 * Allocator for rtio_blocks
+	 */
+	struct rtio_allocator *allocator;
+
+	/**
+	 * Fifo to output blocks to
+	 */
+	struct k_fifo *fifo;
+
+	/** Time in cycles to read before making the block ready
+	 *
+	 * Should be set using something like K_MSEC(5) or K_FOREVER
+	 */
+	u32_t time;
```

Comment:
What is the purpose of this field?

Reply:
In cases where you'd like to meet some time limited capture of data.
Imagine an IMU sensor that does step detection or a touch screen sensor, you'd like to see if there were steps in the last 10 seconds or perhaps touches in the last 10ms, whether or not anything happened. This time is the time limit to capture data before putting the block into the FIFO, empty or not. In the case of K_FOREVER the block would wait to be filled or meet the capture count in the policy.

Reply:
So IMHO `timeout` will be better name. Also, there should be better explanation in the comment.